File: maximal-square/maximal-square.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        
        
        int m = matrix.length; 
        int n = matrix[0].length;
        
        int[][] grid = new int[m][n];
        int max_s = 0;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                grid[i][j] = matrix[i][j] - '0';
                max_s = Math.max(grid[i][j], max_s);
            }
        }
        
        
        
        for(int i = 1; i < m; i++){
            for(int j = 1; j < n; j++){
                if(grid[i][j] == 0)continue;
                
                grid[i][j] = Math.min(Math.min(grid[i-1][j], grid[i][j-1]), grid[i-1][j-1]) + 1;
                max_s = Math.max(grid[i][j], max_s);
            }
        }
        
        return max_s * max_s;
        
        
        
    }
}

File: maximal-square/maximal-square.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        
        
        m = len(matrix)
        n = len(matrix[0])
        max_val = 0
        
        for i in range(n):
            if matrix[0][i] == "1":
                max_val = 1
        
        for i in range(m):
            if matrix[i][0] == "1":
                max_val = 1
        
        
        for i in range(1, m):
            for j in range(1, n):
                
                if matrix[i][j] == "0":
                    continue
                    
                min_val = min(int(matrix[i-1][j-1]), min(int(matrix[i-1][j]), int(matrix[i][j-1])))
                matrix[i][j] = min_val+1
                max_val = max(int(matrix[i][j]), max_val)
        
        return max_val ** 2